src/Lexer/DOMLex.hack

```diff
@@ -9,227 +9,255 @@ use namespace HTMLPurifier\Token;
-        $node_map = $node_map;
-        if ($node_map is null) {
-            throw new \Exception('Node Map should be nonnull, but it is null');
-        }
-        if ($node_map->length === 0) {
-            return dict[];
-        }
-        $array = dict[];
-        foreach ($node_map as $key => $value) {
-            $array[$key] = $value->nodeValue;
-        }
-
-        return $array;
-    }
-
-    //callback function for undoing escaping of stray 
-    public function callbackUndoCommentSubst(string $comment_content, string $closing_marker): string {
-        return '<!--' . Str\replace_every($comment_content, dict['&amp;' => '&', '&lt' => '<']) . $closing_marker;
-    }
-
-    // callback function that entity-izes ampersands in comments so that callbackUndoCommentSubst doesn't clobber them
-    public function callbackArmorCommentEntities(string $comment_content, string $closing_marker): string {
-         return '<!--' . Str\replace('&', '&amp;', $comment_content) . $closing_marker;
-    }
-
-    //wraps an HTML fragment in the necessary HTML
-    protected function wrapHTML(string $html, HTMLPurifier\HTMLPurifier_Config $config, 
-        HTMLPurifier\HTMLPurifier_Context $_context, ?bool $use_div = true) : string {
-        $def = $config->getDefinition('HTML');
-        $ret = '';
-
-        //if it contains a definition and definition doctype, check its public and system strings to append
-        if ($def && ($def->doctype) && ($def->doctype->dtdPublic || $def->doctype->dtdSystem)) {
-            $ret .= '<!DOCTYPE html ';
-            if ($def->doctype->dtdPublic) {
-                $ret .= 'PUBLIC "' . $def->doctype->dtdPublic . '" ';
-            }
-            if ($def->doctype->dtdSystem) {
-                $ret .= '"' . $def->doctype->dtdSystem . '" ';
-            }   
-            $ret .= '>';
-        }
-
-        $ret .= '<html><head>';
-        $ret .= '<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />';
-        // No protection if $html contains a stray </div>!
-        $ret .= '</head><body>';
-        if ($use_div) $ret .= '<div>';
-        $ret .= $html;
-        if ($use_div) $ret .= '</div>';
-        $ret .= '</body></html>';
-        return $ret;
-    }
+	private HTMLPurifier\HTMLPurifier_TokenFactory $factory;
+
+	//set up factory
+	public function __construct() {
+		parent::__construct();
+		$this->factory = new HTMLPurifier\HTMLPurifier_TokenFactory();
+	}
+
+	public function muteErrorHandler(): void {
+	}
+
+	public function tokenizeHTML(
+		string $html,
+		HTMLPurifier\HTMLPurifier_Config $config,
+		HTMLPurifier\HTMLPurifier_Context $context,
+	): vec<HTMLPurifier\HTMLPurifier_Token> {
+		$html = $this->normalize($html, $config, $context);
+		// attempt to armor stray angled brackets that cannot possibly
+		// form tags and thus are probably being used as emoticons
+		if ($config->def->defaults['Core.AggressivelyFixLt']) {
+			while (Str\contains($html, '<!--')) {
```

Comment:
initially was confused by this cause the POC still renders in the browser if you add spaces (like `<! --><iframe srcdoc='<script>alert(1)</script>'>-->`) but it seems like later on we parse it in a way that drops the comments after this `while` cause I can't get the tests to break in the right way :thinking:

Reply:
Hmm interesting... We should look some more into this when we review the regexes